### PR TITLE
Component archive

### PR DESCRIPTION
**Pull request by fabianburth**
**What this PR does / why we need it**:
The possibility to have Component Archives with with uncompressed/unpacked resources in a directory enables more convenient test cases. This is especially useful as it allows the ocm-lib to deal with the already existing test cases in the gardener landscaper. 

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release Notes**:

```feature user

```

### CODE REVIEW

File: pkg/contexts/ocm/repositories/comparch/accessmethod_localfs.go [hunk 1]
```diff
@@ -39,13 +40,31 @@ func (m *localFilesystemBlobAccessMethod) GetKind() string {
...
+	}
+	return accessio.BlobData(m.blobAccess, nil)
 }
 
 func (m *localFilesystemBlobAccessMethod) MimeType() string {
 	return m.spec.MediaType
 }
+
+func (m *localFilesystemBlobAccessMethod) Close() error {
+	return m.blobAccess.Close()
```
Comment (mandelsoft) [inline pkg/contexts/ocm/repositories/comparch/accessmethod_localfs.go:69]: You have to check for nil, first.

And you should introduce a `sync.Mutex` here, to protect the handling of the cached access.

File: pkg/contexts/ocm/repositories/comparch/accessmethod_localfs.go [hunk 2]
```diff
@@ -39,13 +40,31 @@ func (m *localFilesystemBlobAccessMethod) GetKind() string {
...
 func (m *localFilesystemBlobAccessMethod) Get() ([]byte, error) {
-	return accessio.BlobData(m.base.GetBlobData(m.spec.LocalReference))
+	if m.blobAccess == nil {
+		var err error
+		m.blobAccess, err = m.base.GetBlobData(m.spec.LocalReference)
+		if err != nil {
+			return accessio.BlobData(m.blobAccess, err)
+		}
+	}
+	return accessio.BlobData(m.blobAccess, nil)
```
Comment (mandelsoft) [inline pkg/contexts/ocm/repositories/comparch/accessmethod_localfs.go:61]: yes, the original code never closed the data access. This is definitely  the right way to deal with this.: cache the access until the method is closed.

File: pkg/contexts/ocm/repositories/comparch/comparch_test.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	})
+
+	It("component archive with a resource stored in a directory", func() {
+		octx := ocm.DefaultContext()
+		spec := Must(comparch.NewRepositorySpec(accessobj.ACC_WRITABLE, DIR_COMPARCH))
+		repo := Must(spec.Repository(octx, nil))
+		cv := Must(repo.LookupComponentVersion("example.com/root", "1.0.0"))
+		res := Must(cv.GetResourcesByName("root-a"))
+		acc := Must(res[0].AccessMethod())
+		defer acc.Close()
```
Comment (mandelsoft) [inline pkg/contexts/ocm/repositories/comparch/comparch_test.go:76]: you should use the `defer Close(acc)`-like way supported by the testutils. It checks for the err code, also.

File: pkg/contexts/ocm/repositories/comparch/comparch_test.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		bufferB := bytes.NewBuffer(bytesB)
+		r, _ := Must2(compression.AutoDecompress(bufferB))
+		_, _ = Must2(tarutils.ExtractTarToFsWithInfo(mfs, r))
+		Expect(bytesA).To(Equal(bufferB.Bytes()))
+	})
+
+	It("component archive with a resource stored in a directory", func() {
+		octx := ocm.DefaultContext()
+		spec := Must(comparch.NewRepositorySpec(accessobj.ACC_WRITABLE, DIR_COMPARCH))
+		repo := Must(spec.Repository(octx, nil))
```
Comment (mandelsoft) [inline pkg/contexts/ocm/repositories/comparch/comparch_test.go:72]: Close the repo (see below)

File: pkg/contexts/ocm/repositories/comparch/comparch_test.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+		cv := Must(repo.LookupComponentVersion("example.com/root", "1.0.0"))
```
Comment (mandelsoft) [inline pkg/contexts/ocm/repositories/comparch/comparch_test.go:73]: Close the cv (see below)

File: pkg/contexts/ocm/repositories/comparch/comparch_test.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	It("marshal/unmarshal simple", func() {
+		octx := ocm.DefaultContext()
+		spec := Must(comparch.NewRepositorySpec(accessobj.ACC_WRITABLE, TEST_FILEPATH))
+		data := Must(json.Marshal(spec))
+		Expect(string(data)).To(Equal("{\"type\":\"" + comparch.Type + "\",\"filePath\":\"" + TEST_FILEPATH + "\"}"))
+		_ = Must(octx.RepositorySpecForConfig(data, runtime.DefaultJSONEncoding)).(*comparch.RepositorySpec)
+		// spec will not equal r as the filesystem cannot be serialized
+	})
+
+	It("component archive with resource stored as tar", func() {
```
Comment (mandelsoft) [inline pkg/contexts/ocm/repositories/comparch/comparch_test.go:43]: see comments for test below

File: pkg/contexts/ocm/repositories/comparch/testdata/directory/blobs/root/blueprint.yaml [hunk 7]
```diff
@@ -0,0 +1,44 @@
+# SPDX-FileCopyrightText: 2020 SAP SE or an SAP affiliate company and Gardener contributors
```
Comment (mandelsoft) [inline pkg/contexts/ocm/repositories/comparch/testdata/directory/blobs/root/blueprint.yaml:1]: may be you should just use some data here, and not a real blueprint.

**Review by mandelsoft — CHANGES_REQUESTED** [review]

File: pkg/contexts/ocm/plugin/plugin_test.go [hunk 8]
```diff
@@ -9,14 +9,14 @@ package plugin_test
 import (
 	. "github.com/onsi/ginkgo/v2"
 	. "github.com/onsi/gomega"
-	"github.com/open-component-model/ocm/pkg/contexts/ocm/cpi"
```
[the remark this answers is not in the document]
Comment (fabianburth) [reply pkg/contexts/ocm/plugin/plugin_test.go:12]: Yes, but conveniently running `make prepare` (or `make format`) before pushing without affecting files other than the ones edited in the specific PR would require everyone to this each time before merging their PR.

File: pkg/common/accessio/opts.go [hunk 9]
```diff
@@ -43,7 +43,7 @@ type Options interface {
 
 type StandardOptions struct {
 	// FilePath is the path of the repository base in the filesystem
-	FileFormat *FileFormat `json:"fileFormat"`
+	FileFormat *FileFormat `json:"fileFormat,omitempty"`
```
[the remark this answers is not in the document]
Comment (fabianburth) [reply pkg/common/accessio/opts.go:46]: If the file format is not set, the format is automatically detected during the `spec.Repository()`call.

File: docs/pluginreference/plugin_accessmethod.md [hunk 10]
```diff
@@ -28,6 +28,7 @@ described by an access method descriptor ([plugin descriptor](plugin_descriptor.
 
 * [plugin accessmethod <b>compose</b>](plugin_accessmethod_compose.md)	 &mdash; compose access specification from options and base specification
 * [plugin accessmethod <b>get</b>](plugin_accessmethod_get.md)	 &mdash; get blob
+* [plugin accessmethod <b>identity</b>](plugin_accessmethod_identity.md)	 &mdash; get blob identity
```
[the remark this answers is not in the document]
Comment (fabianburth) [reply docs/pluginreference/plugin_accessmethod.md:31]: It appears so (was generated during `make prepare` - we should probably make it a habit running it for every PR).

**Review by mandelsoft — APPROVED** [review]

**Review by mandelsoft — APPROVED** [review]

**Pull request merged**